File: deck/seed/code/native/kotlin/runtime/http.kt
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest

object http {
    // `header` is a map of name to value and may be empty. Written to match
    // the node runtime, and NOT exercised here: this repository builds and
    // tests the node target only.
    suspend fun request(method: String, url: String, body: String, header: Map<String, String>): HttpResponse {
        return try {
            val builder = HttpRequest.newBuilder().uri(URI.create(url))
            for ((name, value) in header) builder.header(name, value)
            if (method == "GET") builder.GET() else builder.method(method, HttpRequest.BodyPublishers.ofString(body))
            val resp = HttpClient.newHttpClient().send(builder.build(), java.net.http.HttpResponse.BodyHandlers.ofString())
            HttpResponse(resp.statusCode().toLong(), resp.body())
        } catch (e: Exception) {
            HttpResponse(0L, "")
        }
    }
}

File: deck/seed/code/native/swift/runtime/http.swift
import Foundation

enum http {
    // `header` is a map of name to value and may be empty. Written to match
    // the node runtime, and NOT exercised here: this repository builds and
    // tests the node target only.
    static func request(_ method: String, _ url: String, _ body: String, _ header: SeedMap<String, String>) async -> HttpResponse {
        let header = header.data
        guard let u = URL(string: url) else { return HttpResponse(status: 0, body: "") }
        var req = URLRequest(url: u)
        req.httpMethod = method
        for (name, value) in header { req.setValue(value, forHTTPHeaderField: name) }
        if !body.isEmpty { req.httpBody = body.data(using: .utf8) }
        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            return HttpResponse(status: status, body: String(data: data, encoding: .utf8) ?? "")
        } catch {
            return HttpResponse(status: 0, body: "")
        }
    }
}
